internal/controllers/machine/machine_controller.go

```diff
@@ -383,6 +401,25 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, s *scope) (ctrl.Result
 	cluster := s.cluster
 	m := s.machine
 
+	// Check reconcileDeleteCache to ensure we won't run reconcileDelete too frequently.
+	if cacheEntry, ok := r.reconcileDeleteCache.Has(cache.NewReconcileEntryKey(s.machine)); ok {
+		if requeueAfter, requeue := cacheEntry.ShouldRequeue(time.Now()); requeue {
+			return ctrl.Result{RequeueAfter: requeueAfter}, nil
+		}
+	}
+
+	s.reconcileDeleteExecuted = true
+
+	// Add entry to the reconcileDeleteCache so we won't run reconcileDelete more than once per second.
+	// Under certain circumstances the ReconcileAfter time will be set to a later time, e.g. when we're waiting
+	// for Pods to terminate or volumes to detach.
+	// This is done to ensure we're not spamming the workload cluster API server.
+	r.reconcileDeleteCache.Add(cache.NewReconcileEntry(s.machine, time.Now().Add(1*time.Second)))
+
+	// Set "fallback" reason and message. This is used if we don't set a more specific reason and message below.
+	s.deletingReason = clusterv1.MachineDeletingDeletionTimestampSetV1Beta2Reason
```

Comment:
q: what are the cases where we don't set a more specific reason?
I just want to double check that there are no conditions were deletions can be stuck without users getting additional info

Reply:
I'm mostly doing this because it's very hard to reason through all the branches below for all edge cases

Reply:
Audited the cases:
```go
return ctrl.Result{}, errors.Wrapf(err, "failed to check if Kubernetes Node deletion is allowed")
return ctrl.Result{}, err (l.460)
return ctrl.Result{}, errors.Wrap(err, "failed to patch Machine") (l.478)
return result, nil (l.491) because of return ctrl.Result{RequeueAfter: time.Minute}, nil (l.765)
return ctrl.Result{}, err (l.549, reconcileDeleteInfrastructure)
return ctrl.Result{}, err (l.561, reconcileDeleteBootstrap)
```

So I'll add deletingReason & deletingMessage for all of these as well.

Please double check, I'm absolutely not sure if I missed some cases

Reply:
I'll defer this to you
I'm ok if we skip the less probable cases (where e.g. we read only stuff from the cache, or we perform simple operations on the management cluster)

Reply:
I added reasons & message for all of these cases. It would be just good if someone else also goes through all branches to ensure I didn't miss any. It's non-trivial :)

Reply:
I double checked the code paths, it seems ok to me